openff/toolkit/topology/molecule.py

```diff
@@ -5785,9 +5790,28 @@ def sort_hierarchy_elements(self):
         their identifiers.
         """
 
-        # hard-code the sort_func value here, since it's hard to serialize safely
-        def sort_func(x):
-            return version.parse(".".join([str(i) for i in x.identifier]))
+        def sort_func(x: HierarchyElement) -> version.Version:
+            """Sort by int-like tags, using version.Version as a quick hack."""
+            tag = ""
+            for i in x.identifier:
+                if isinstance(i, int):
+                    _str_i = str(i)
+                elif isinstance(i, str):
+                    try:
+                        int(i)
+                    except ValueError:
+                        continue
+                    _str_i = i
+                else:
+                    raise Exception(f"Unexpected type {type(i)} as identifier")
```

Comment:
This could maybe do without existing, but non-`Union[int, str]` sneaking into `HierarchyElement.identifier` makes me nervous and I haven't explored how possible it might be with sloppy inputs